gnvim.1.in
.TH GNVIM 1 gnvim\-@VERSION@
.SH NAME
gnvim \- nvim terminal wrapper
.SH SYNOPSIS
.B gnvim
.RB [ options ]
.IR file
.SH DESCRIPTION
The program gnvim is a wrapper script to allow gvim like functionality for neovim along a terminal
emulator, ie: when launched with or without files it will search for an instance of 'gnvim' by
checking on the runfile, the runfile is typically located at
.B /var/run/user/${UID}/${class}/${class}.lock
after which gnvim either creates a new terminal window via the
.B x-terminal-emulator
command or simply hands out the file to the running instance that posses a terminal

The actual handling of the file and arguments is done by an instance of neovim, either by itself
listening to the p_file located at:
.B /var/run/user/${UID}/${class}/${class}.pipe
OR an instance of neovim running inside a tmux session started by
.B tmux-nvim
(1)

The default terminal emulator to handle the terminal window is the
.B x-terminal-emulator
interface, a debianism that may be a wrapper or a symlink to a terminal emulator binary, this is
because for the intended functionality of this script the terminal emulator should be compatible
with the following xterm(1) options: -title, -name, -e

Proper handling of the options and a compatible behaviour is expected however not all terminal are
compatible out of the box, debian does by policy provide wrappers which then get symlinked onto the
x-terminal-emulator, an example would be the gnome-terminal.wrapper from:
.IR https://sources.debian.org/src/gnome-terminal/3.56.2-2/debian/gnome-terminal.wrapper

To consult which terminal emulators in debian currently provide the x-terminal-emulator interface
either by natively supporting the xterm options or through a wrapper check in:
.IR https://packages.debian.org/sid/x-terminal-emulator

If your preferred terminal emulator does not appear in the list you can provide your own wrapper
script and either link it into x-terminal-emulator or simply specify that script as the terminal
command to call inside the configuration file.

.SH OPTIONS
.TP
.B "help, \-h, \-\-help"
Print brief usage information to standard output and exit.
.TP
.B "class, \-c, \-\-class CLASS"
Run instance with class CLASS, all the variables ${class} mentioned above become this name.
.SH CONFIGURATION
.P
The configuration of gnvim is located at:
.B ${XDG_CONFIG_HOME:-${HOME}/.config}/gnvim/configrc

.P
To see the location of the configuration file and generating it if running for the first time, run
gnvim with the help flag

.P
The configuration defines the following options:

.TP
.B
term_cmd
the terminal command to use, should be able to support xterm's "-name", "-title" and "-e" options,
by default the debianism
.B x-terminal-emulator
is used to ensure support of these options within debian and derivates.

.TP
.B
nv_cmd
the neovim command to use, can be set to any custom wrapper command the user chooses.

default
.B /usr/bin/nvim

.TP
.B
opencmd
the open command used internally by neovim to open a file if not using tmux,
supported values: edit, tabedit, split, vsplit, drop

default
.B edit

.TP
.B
class
the class for the gnvim instance.

default
.B gnvim

.TP
.B
no_tmux
sets gnvim to not use tmux sessioning and instead manage the sessions with neovim's built in tabs,
by default this is empty to use tmux-nvim

.SH AUTHOR
.EX
eylles                 <ed.ylles1997 at gmail.com>
.EE
.SH HOMEPAGE
.TP
Source code:
.EE
https://github.com/eylles/xterm-nvim
.EX


gnvim.sh
#!/bin/sh

term_cmd="x-terminal-emulator"
no_tmux=""
nv_cmd="/usr/bin/nvim"
class="gnvim"
opencmd="edit"

config_dir="${XDG_CONFIG_HOME:-${HOME}/.config}/gnvim"
config_file="${config_dir}/configrc"

if [ ! -f "$config_file" ]; then
    if [ ! -d "$config_dir" ]; then
        mkdir -p "$config_dir"
    fi
    cat << __HEREDOC__ > "$config_file"
# vim: ft=sh
# gnvim wrapper config file

# terminal command
term_cmd="${term_cmd}"

# neovim command
nv_cmd="${nv_cmd}"

# open command
# used internally by neovim to open a file if not using tmux
# supported values: edit, tabedit, split, vsplit, drop
opencmd="${opencmd}"

# gnvim class
class="${class}"

# do not use tmux for regular operation
# setting this to anything will take effect
no_tmux="${no_tmux}"
__HEREDOC__
fi

# loading the config here means the default config is already present and the user is not able to
# overwrite anything defined after this point.
. "$config_file"

myname="${0##*/}"

# cuz $UID is not POSIX ¯\_(ツ)_/¯
# but it may be defined in the environment
if [ -z "$UID" ]; then
  # result of: id -u $USER
  UserID=$(id -u "$USER")
else
  UserID="$UID"
fi

case "$1" in
    class|-c|--class)
        shift
        class="$1"
        shift
        ;;
    help|-h|--help)
        printf '%s\n\t%s\n' "${myname}:" \
                            "nvim terminal wrapper"
        printf '%s\n\t%s\n' "Version:" "@VERSION@"
        printf '%s\n\t%s\n' "Usage:" \
                            "${myname} -h | [ --class NAME ] [ nvim flags and files ]"
        printf '%s\n\t%s\n' "    -h" \
                            "print out this help message, 'help' and '--help' are supported too"
        printf '%s\n\t%s\n\t%s%s\n' \
                            "    --class NAME" \
                            "sets terminal '-name' and '-T' flags to NAME, sets tmux-nvim" \
                            "--session-id to NAME, sets runfile as " \
                            "'/var/run/user/ID/NAME/NAME.lock'"
        printf '\n'
        printf '%s\n\t%s\n' "Config:" "$config_file"
        exit 0
        ;;
esac

run_dir="${XDG_RUNTIME_DIR:-/run/user/${UserID}}/${class}"
run_file="${run_dir}/${class}.lock"
p_file="${run_dir}/${class}.pipe"

# unix command line like booleans

# Type: int
# value: 0
b_true=0
# Type: int
# value: 1
b_false=1

# check if directory is empty
# return type: int
# return values:
#     empty     b_true
#     non empty b_false
# taken from:
#     https://superuser.com/questions/352289/bash-scripting-test-for-empty-directory/830902#830902
is_empty() {
    test -e "$1/"* 2>/dev/null
    case $? in
        1)   return "$b_true" ;;
        *)   return "$b_false" ;;
    esac
}

cleanup() {
    if [ -r "$run_file" ]; then
        rm "$run_file" 2>/dev/null
    fi
    if is_empty "$run_dir"; then
        rmdir "$run_dir" 2>/dev/null
    fi
}

trap cleanup EXIT

if [ -f "$run_file" ]; then
    hasterm="$b_true"
    date +"[%F %T]: ${myname} called." >> "$run_file"
else
    hasterm="$b_false"
    if [ ! -d "$run_dir" ]; then
        mkdir -p "$run_dir"
    fi
    date +"[%F %T]: ${myname} called." > "$run_file"
fi

if [ ! -f "$run_file" ]; then
    exit 1
fi
if [ "$hasterm" -eq "$b_true" ]; then
    if [ -z "$no_tmux" ]; then
        exec tmux-nvim --session-id "$class" "$@"
    else
        case "$opencmd" in
            edit|tabedit|split|vsplit|drop)
                : # do nothing, we just accept those values
                ;;
            *) # fallback to default if not a supported command
                opencmd="edit"
                ;;
        esac
        exec nvim --server "$p_file" --remote-send "<C-\\><C-N>:${opencmd} ${*}<CR>"
    fi
else
    if [ -z "$no_tmux" ]; then
        $term_cmd -name "$class" -title "$class" -e tmux-nvim --session-id "$class" "$@"
    else
        $term_cmd -name "$class" -title "$class" -e $nv_cmd --listen "$p_file" "$@"
    fi
fi
